shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/PatternsByStop.kt

```diff
@@ -99,4 +101,143 @@ data class PatternsByStop(
+        // will split the grouped direction up into as many headsign rows as there are headsigns.
+        fun resolveRealtimePatternForDirection(
+            staticData: NearbyStaticData.StaticPatterns.ByDirection,
+            upcomingTripsMap: UpcomingTripsMap?,
+            parentStopId: String,
+            alerts: Collection<Alert>?,
+            hasSchedulesTodayByPattern: Map<String, Boolean>?,
+            allDataLoaded: Boolean,
+        ): List<RealtimePatterns> {
+            val typicalPatternsByRoute =
+                staticData.patterns
+                    .filter { it.typicality == RoutePattern.Typicality.Typical }
+                    .groupBy({ it.routeId }, { it.id })
+
+            // If data hasn't loaded, or there are more than one typical routes in this direction,
+            // we never want to split into individual headsign rows.
+            if (!allDataLoaded || typicalPatternsByRoute.size > 1) {
+                return listOf(
+                    RealtimePatterns.ByDirection(
+                        staticData,
+                        upcomingTripsMap,
+                        parentStopId,
+                        alerts,
+                        hasSchedulesTodayByPattern,
+                        allDataLoaded
+                    )
+                )
+            }
+
+            val patternsById = staticData.patterns.associateBy { it.id }
+            val tripsByPattern =
+                upcomingTripsMap
+                    .orEmpty()
+                    .mapNotNull {
+                        when (val key = it.key) {
+                            is RealtimePatterns.UpcomingTripKey.ByRoutePattern ->
+                                if (
+                                    key.routePatternId != null &&
+                                        patternsById.containsKey(key.routePatternId) &&
+                                        key.parentStopId == parentStopId
+                                ) {
+                                    key.routePatternId to it.value
+                                } else {
+                                    null
+                                }
+                            else -> null
+                        }
+                    }
+                    .toMap()
+            val upcomingPatternsByRouteAndHeadsign =
+                tripsByPattern.values
+                    .flatten()
+                    .groupBy({ it.trip.routeId }, { it.trip.headsign to it.trip.routePatternId })
+                    .mapValues { routeEntry ->
+                        routeEntry.value.groupBy({ it.first }, { it.second })
+                    }
+
+            val predictedPatternsByRoute =
+                staticData.patterns
+                    .filter {
+                        tripsByPattern[it.id]?.any { trip ->
+                            trip.prediction != null && trip.vehicle?.tripId == trip.trip.id
```

Comment:
question: why check the vehicle's trip ID here? Wouldn't we still show the prediction even if the vehicle is on another trip?

Reply:
Hmm, maybe it's a bad idea. I wanted to prevent cases where we would be displaying the grouping without actually displaying the atypical trip. So like, if there was a prediction for a C which is currently approaching Medford Tufts to start a new trip, but two E trains in front of it heading west, we would still show the group even though only the E trains are shown in the upcoming predictions. But maybe that's fine and this is too cautious and will hide predictions when we don't want it to.

Reply:
I removed this check, I think it'll be okay to briefly show groups sometimes even when it's not totally clear why.